pkg/management/postgres/wal.go

```diff
@@ -28,63 +28,41 @@ import (
 	"github.com/cloudnative-pg/cloudnative-pg/pkg/resources"
 )
 
-type dbProvider func() (*sql.DB, error)
+var errNoWalArchivePresent = errors.New("no wal-archive present")
 
-type walArchiveBootstrapper struct {
-	firstWalArchiveTriggered bool
-	backoff                  *wait.Backoff
-	dbProviderFunc           dbProvider
-}
-
-func newWalArchiveBootstrapper() *walArchiveBootstrapper {
-	return &walArchiveBootstrapper{}
+type walArchiveAnalyzer struct {
+	dbFactory func() (*sql.DB, error)
+	backoff   wait.Backoff
 }
 
-func (w *walArchiveBootstrapper) withTimeout(backoff *wait.Backoff) *walArchiveBootstrapper {
-	w.backoff = backoff
-	return w
-}
-
-func (w *walArchiveBootstrapper) withDBProvider(provider dbProvider) *walArchiveBootstrapper {
-	w.dbProviderFunc = provider
-	return w
+func newWalArchiveAnalyzer(backoff wait.Backoff, primaryConnInfo string) *walArchiveAnalyzer {
+	return &walArchiveAnalyzer{
+		dbFactory: func() (*sql.DB, error) {
+			db, openErr := sql.Open(
+				"pgx",
+				fmt.Sprintf("%s dbname=%s", primaryConnInfo, "postgres"),
+			)
+			if openErr != nil {
+				log.Error(openErr, "can not open postgres database")
+				return nil, openErr
+			}
+			return db, nil
+		},
+		backoff: backoff,
+	}
 }
 
-func (w *walArchiveBootstrapper) withInstanceDBProvider(instance *Instance) *walArchiveBootstrapper {
-	return w.withDBProvider(func() (*sql.DB, error) {
-		db, openErr := sql.Open(
-			"pgx",
-			fmt.Sprintf("%s dbname=%s", instance.GetPrimaryConnInfo(), "postgres"),
-		)
-		if openErr != nil {
-			log.Error(openErr, "can not open postgres database")
-			return nil, openErr
+func (w *walArchiveAnalyzer) execute() error {
+	return retry.OnError(w.backoff, resources.RetryAlways, func() error {
+		db, err := w.dbFactory()
+		if err != nil {
+			return err
 		}
-		return db, nil
+		return w.mustHaveFirstWalArchived(db)
 	})
 }
```

Comment:
this connection is never closed, is that intentional? @armru

Reply:
yes this was intentional in the old implementation, but I've streamlined the logic further